### PR TITLE
Local_library

### PR DESCRIPTION
**Pull request by aniketrathi**
## Description
- Home
#### Description
GET/
Show the counts of books, book-instances, genre, authors.
Response:
200
```
index.ejs
```
- All authors
#### Description
GET/authors
Show all authors.
Response:
200
```
author-list.ejs
```
- All books
#### Description
GET/books
Show all books.
Response:
200
```
book-list.ejs
```
- All bookinstances
#### Description
GET/bookinstances
Show all book-instances
Response:
200
```
bookinstance-list.ejs
``` 
- All genres
#### Description
GET/genres
Show all genres
Response:
200
```
genre-list.ejs
``` 
- Author details
#### Description
GET/author/:id
Show the name, date of birth, date of death and books of the author .
Response:
200
```
author-detail.ejs
``` 
- Book details
#### Description
GET/book/:id
Show the title, summary, author, genre and book-instances of the book .
Response:
200
```
book-detail.ejs
``` 
- Book-instance details
#### Description
GET/bookinstance/:id
Show the Id, status, imprint, book-title of the that book-instance.
Response:
200
```
bookinstance-detail.ejs
``` 
- Genre details
#### Description
GET/genre/:id
Show the name, book's title and book's summary of that genre .
Response:
200
```
genre-detail.ejs
``` 
- Create new author
#### Description
GET/author/create
Show the form to create new author.
Response:
200
```
author-form.ejs
``` 
POST/author/create
Add details of the author
Request Body:
```
{
 first_name: 'author's first name',
 family_name: 'author's family name',
 date_of_birth: 'author's dob',
 date_of_death: 'author's dod'
}
```
Response:
200
```
redirect to author-detail.ejs
```
- Create new book
#### Description
GET/book/create
Show the form to create new book.
Response:
200
```
 book-form.ejs
``` 
POST/book/create
Add details of the book.
Request Body:
```
{
 title: 'book's title',
 author: 'book's author',
 summary: 'book's summary',
 isbn: 'book's isbn',
 genre: 'book's genre' 
}
```
Response:
200
```
redirect to book-detail.ejs
```
- Create new book instance
#### Description
GET/bookinstance/create
Show the form to create new book instance.
Response:
200
```
 bookinstance-form.ejs
``` 
POST/bookinstance/create
Add details of the bookinstance.
Request Body:
```
{
 book: 'book's title',
imprint: 'bookinstance's imprint',
due_back: 'book's due back date',
status: 'book's status',
}
```
Response:
200
```
redirect to bookinstance-detail.ejs
```
- Create new genre
#### Description
GET/genre/create
Show the form to create new genre.
Response:
200
```
 genre-form.ejs
``` 
POST/genre/create
Add details of the book genre.
Request Body:
```
{
 name: 'genre name' 
}
```
Response:
200
```
redirect to genre-detail.ejs
```
- Delete author
#### Description
GET/author/:id/delete
Show name , dob and dod of author with a message which say first delete associated book of it. 
Response:
200
```
 author-delete.ejs
``` 
POST/author/:id/delete
Request Body:
```
{
 id:'author._id' 
}
```
Response:
200
```
redirect to author-list.ejs
```
- Delete book
#### Description
GET/book/:id/delete
Show title of book with a message which say first delete associated bookinstances of it. 
Response:
200
```
 book-delete.ejs
``` 
POST/book/:id/delete
Request Body:
```
{
 id:'book._id' 
}
```
Response:
200
```
redirect to book-list.ejs
```
- Delete bookinstance
#### Description
GET/bookinstance/:id/delete
Show the id of book instance.
Response:
200
```
 book-delete.ejs
``` 
POST/bookinstance/:id/delete
Request Body:
```
{
 id:'book._id' 
}
```
Response:
200
```
redirect to bookinstance-list.ejs
```
- Delete genre
#### Description
GET/genre/:id/delete
Show name of genre with a message which say first delete associated books of it. 
Response:
200
```
 genre-delete.ejs
``` 
POST/genre/:id/delete
Request Body:
```
{
 id:'book._id' 
}
```
Response:
200
```
redirect to genre-list.ejs
```
- Update author
#### Description
GET/author/:id/update
Show form of the author with filled entries that can be alter.
Response:
200
```
 author-form.ejs
``` 
POST/author/:id/update
Request Body:
```
{
  first_name: 'author's first name',
 family_name: 'author's family name',
 date_of_birth: 'author's dob',
 date_of_death: 'author's dod'
}
```
Response:
200
```
redirect to author-details.ejs
```
- Update a book
#### Description
GET/book/:id/update
Show the form of book with filled entries which can be alter.
Response:
200
```
 book-form.ejs
``` 
POST/book/:id/update
Add details of the book.
Request Body:
```
{
 title: 'book's title',
 author: 'book's author',
 summary: 'book's summary',
 isbn: 'book's isbn',
 genre: 'book's genre' 
}
```
Response:
200
```
redirect to book-detail.ejs
```
- Update book instance
#### Description
GET/bookinstance/:id/update
Show the form of book instance with filled entries which can be alter.
Response:
200
```
 bookinstance-form.ejs
``` 
POST/bookinstance/:id/update
Add details of the bookinstance.
Request Body:
```
{
 book: 'book's title',
imprint: 'bookinstance's imprint',
due_back: 'book's due back date',
status: 'book's status',
}
```
Response:
200
```
redirect to bookinstance-detail.ejs
```
- Update genre
#### Description
GET/genre/:id/update
Show the form of genre with filled entries which can be alter.
Response:
200
```
 genre-form.ejs
``` 
POST/genre/:id/update
Add details of the book genre.
Request Body:
```
{
 name: 'genre name' 
}
```
Response:
200
```
redirect to genre-detail.ejs
```
## Database schema changes
_None_

### Automated test cases added
- _None_

### Manual test cases run
- _None_

### CODE REVIEW

**Comment by GaurisankarJ** [comment]
Please add the route next to the method in your description.

Eg:
POST /account, here `POST` is the method and `/account` is the route.

File: app.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
+var cookieParser = require("cookie-parser");
```
Comment (GaurisankarJ) [inline app.js]: Please use `const` for imports rather than `var`.

File: app.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+var indexRouter = require("./routes/index");
+var usersRouter = require("./routes/users");
+var catalogRouter = require("./routes/catalog");
+
+var app = express();
+
+env.config();
+
+const dev_db_url =
+`mongodb+srv://Aniket:${process.env.password}@cluster0.9voqa.mongodb.net/${process.env.dbname}?retryWrites=true&w=majority`;
```
Comment (GaurisankarJ) [inline app.js]: Please store dev db url in the env file too.

File: controllers/author-controller.js [hunk 3]
```diff
@@ -0,0 +1,210 @@
+var Author = require("../models/author");
```
Comment (GaurisankarJ) [inline controllers/author-controller.js]: Use `const` please.

File: controllers/author-controller.js [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+      //Successful, so render
+      res.render("author-list", {
+        title: "Author List",
+        author_list: list_authors,
+      });
+    });
+};
+
+// Display detail page for a specific Author.
+exports.author_detail = function (req, res) {
```
Comment (GaurisankarJ) [inline controllers/author-controller.js:25]: `const { id } = req.params;`

Please extract and store params and query fields in variables in the beginning, they may be used multiple times in the function body.

File: controllers/author-controller.js [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+};
+
+// Display Author create form on GET.
+exports.author_create_get = function (req, res) {
+  res.render("author-form", { title: "Create Author" });
+};
+
+// Handle Author create on POST.
+exports.author_create_post = [
+  // Validate and sanitise fields.
```
Comment (GaurisankarJ) [inline controllers/author-controller.js:63]: Please move the validation part to a new file, make a validators folder in the root.

Comment (GaurisankarJ) [reply controllers/author-controller.js:63]: Run the validator before the `author_create_post` is called.

File: controllers/author-controller.js [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+        author: req.body,
+        errors: errors.array(),
+      });
+      return;
+    } else {
+      // Data from form is valid.
+
+      // Create an Author object with escaped and trimmed data.
+      var author = new Author({
+        first_name: req.body.first_name,
```
Comment (GaurisankarJ) [inline controllers/author-controller.js]: Please extract and store values from the body at the beginning of the function.

File: controllers/author-controller.js [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+// Handle Author update on POST.
+exports.author_update_post = function (req, res) {
+  Author.findByIdAndUpdate(req.params.id, req.body).exec(function (
+    err,
+    updated_author
+  ) {
+    if (err) return next(err);
+    else {
+      console.log(updated_author);
```
Comment (GaurisankarJ) [inline controllers/author-controller.js]: Remove the `console.log`.

File: controllers/book-controller.js [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+  // Convert the genre to an array.
+  (req, res, next) => {
+    if (!(req.body.genre instanceof Array)) {
+      if (typeof req.body.genre === "undefined") req.body.genre = [];
+      else req.body.genre = new Array(req.body.genre);
+    }
+    next();
+  },
+
+  // Validate and sanitise fields.
```
Comment (GaurisankarJ) [inline controllers/book-controller.js:119]: Add this to a new validator file for the particular route.

File: controllers/bookinstance-controller.js [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+      title: "Create BookInstance",
+      book_list: books,
+    });
+  });
+};
+
+// Handle BookInstance create on POST.
+exports.bookinstance_create_post = [
+  // Validate and sanitise fields.
+  body("book", "Book must be specified").trim().isLength({ min: 1 }).escape(),
```
Comment (GaurisankarJ) [inline controllers/bookinstance-controller.js:64]: Move to a new validator file.

File: controllers/bookinstance-controller.js [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+// Handle bookinstance update on POST.
+exports.bookinstance_update_post = function (req, res) {
+  BookInstance.findByIdAndUpdate(req.params.id, req.body).exec(function (
+    err,
+    updated_bookinstance
+  ) {
+    if (err) return next(err);
+    else {
+      console.log(updated_bookinstance);
```
Comment (GaurisankarJ) [inline controllers/bookinstance-controller.js]: Remove `console.log`.

File: populatedb.js [hunk 11]
```diff
@@ -0,0 +1,328 @@
+#! /usr/bin/env node
```
Comment (GaurisankarJ) [inline populatedb.js:1]: Is this boilerplate code?

File: routes/users.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
+var express = require("express");
```
Comment (GaurisankarJ) [inline routes/users.js]: Is this required?

File: app.js (hunk 1 above)
Comment (aniketrathi) [reply app.js]: it is a part of boilerplate, Bdw i will change it.

File: app.js (hunk 2 above)
Comment (aniketrathi) [reply app.js]: ok

File: controllers/author-controller.js (hunk 3 above)
Comment (aniketrathi) [reply controllers/author-controller.js]: ok

File: controllers/author-controller.js (hunk 5 above)
Comment (aniketrathi) [reply controllers/author-controller.js:63]: ok

File: populatedb.js (hunk 11 above)
Comment (aniketrathi) [reply populatedb.js:1]: yes

File: app.js (hunk 2 above)
Comment (GaurisankarJ) [reply app.js]: Have you made the change? I do not see it here.

File: controllers/author-controller.js (hunk 4 above)
Comment (GaurisankarJ) [reply controllers/author-controller.js:25]: Please make these changes through out from now for all `params`, `query` and `body` of the req.

File: controllers/author-controller.js (hunk 6 above)
Comment (GaurisankarJ) [reply controllers/author-controller.js]: Please make these changes through out from now for all params, query and body of the req.

File: controllers/author-controller.js (hunk 7 above)
Comment (GaurisankarJ) [reply controllers/author-controller.js]: Please remove.

File: controllers/bookinstance-controller.js (hunk 10 above)
Comment (GaurisankarJ) [reply controllers/bookinstance-controller.js]: Please remove.

File: routes/users.js (hunk 12 above)
Comment (GaurisankarJ) [reply routes/users.js]: Please remove file if not required.

Comment (aniketrathi) [reply routes/users.js]: ohh sorry its just for practice i will do it in no second.

File: app.js (hunk 2 above)
Comment (aniketrathi) [reply app.js]: yes i had store in the env file

File: controllers/author-controller.js (hunk 7 above)
Comment (aniketrathi) [reply controllers/author-controller.js]: i already removed it sir

File: controllers/bookinstance-controller.js (hunk 10 above)
Comment (aniketrathi) [reply controllers/bookinstance-controller.js]: already removed sir

File: controllers/author-controller.js (hunk 6 above)
Comment (aniketrathi) [reply controllers/author-controller.js]: already done sir

File: app.js [hunk 13]
```diff
@@ -7,14 +7,13 @@ const mongoose = require('mongoose');
...
 const indexRouter = require("./routes/index");
-const usersRouter = require("./routes/users");
 const catalogRouter = require("./routes/catalog");
 
 const app = express();
 
 env.config();
 
-const mongoDB = process.env.MONGODB_URI || process.env.dev_db_url;
+const mongoDB = process.env.dev_db_url;
```
Comment (GaurisankarJ) [inline app.js]: `DEV_DB_URL` use caps with env variables please.

**Review by GaurisankarJ — APPROVED** [review]

**Pull request merged**